compiler/rustc_trait_selection/src/traits/engine.rs

```diff
@@ -27,24 +27,36 @@ use rustc_session::config::TraitSolver;
 use rustc_span::Span;
 
 pub trait TraitEngineExt<'tcx> {
-    fn new(tcx: TyCtxt<'tcx>) -> Box<Self>;
-    fn new_in_snapshot(tcx: TyCtxt<'tcx>) -> Box<Self>;
+    fn new(infcx: &InferCtxt<'tcx>) -> Box<Self>;
+    fn new_in_snapshot(infcx: &InferCtxt<'tcx>) -> Box<Self>;
 }
 
 impl<'tcx> TraitEngineExt<'tcx> for dyn TraitEngine<'tcx> {
-    fn new(tcx: TyCtxt<'tcx>) -> Box<Self> {
-        match tcx.sess.opts.unstable_opts.trait_solver {
+    fn new(infcx: &InferCtxt<'tcx>) -> Box<Self> {
+        match infcx.tcx.sess.opts.unstable_opts.trait_solver {
             TraitSolver::Classic => Box::new(FulfillmentContext::new()),
-            TraitSolver::NextCoherence => Box::new(FulfillmentContext::new()),
+            TraitSolver::NextCoherence => {
+                if infcx.use_next_trait_solver() {
+                    Box::new(NextFulfillmentCtxt::new())
+                } else {
+                    Box::new(FulfillmentContext::new())
+                }
+            }
             TraitSolver::Chalk => Box::new(ChalkFulfillmentContext::new()),
             TraitSolver::Next => Box::new(NextFulfillmentCtxt::new()),
```

Comment:
can you make this an exhaustive match on `(infcx.tcx.sess.opts.unstable_opts.trait_solver, infcx.use_next_trait_solver())` and ICE if it is set with `Classic` or not set with `Next`?